File: kotlin/src/main/java/firstAttempt/graphValidTree/0261-graph-valid-tree.java
package firstAttempt.graphValidTree;

import java.util.*;

record JPair<A, B>(A first, B second) {}

class Solution {
    public boolean validTree(int n, int[][] edges) {
        var graph = buildGraph(n, edges);

        var visited = new HashSet<Integer>();
        var hasCycle = hasCycle(0, graph, visited);

        return !hasCycle && visited.size() == n;
    }

    private Map<Integer, List<Integer>> buildGraph(int n, int[][] edges) {
        var graph = new HashMap<Integer, List<Integer>>();

        for (var edge : edges) {
            var from = edge[0];
            var to = edge[1];

            var fromNodes = graph.getOrDefault(from, new ArrayList<>());
            fromNodes.add(to);
            graph.put(from, fromNodes);

            var toNodes = graph.getOrDefault(to, new ArrayList<>());
            toNodes.add(from);
            graph.put(to, toNodes);
        }

        return graph;
    }

    private boolean hasCycle(int startNode,
                             Map<Integer, List<Integer>> graph,
                             Set<Integer> visited) {
        var stack = new LinkedList<JPair<Integer, Integer>>();
        stack.add(new JPair<>(null, startNode));

        var nodeToPreviousNode = new HashMap<Integer, Integer>();

        while (!stack.isEmpty()) {
            var pair = stack.pop();
            var from = pair.first();
            var node = pair.second();

            var previousFrom = nodeToPreviousNode.get(node);
            if (nodeToPreviousNode.containsKey(node) && !Objects.equals(previousFrom, from)) {
                return true;
            }

            if (visited.contains(node)) {
                continue;
            }

            if (from != null) nodeToPreviousNode.put(node, from);
            visited.add(node);

            var neighboringNodes = graph.getOrDefault(node, List.of());
            for (var neighbourNode : neighboringNodes) {
                if (neighbourNode.equals(node)) {
                    return true;
                }

                if (!visited.contains(neighbourNode)) {
                    stack.push(new JPair<>(node, neighbourNode));
                }
            }
        }

        return false;
    }
}

File: kotlin/src/main/java/secondAttempt/graphValidTree/0261-graph-valid-tree.kt
package secondAttempt.graphValidTree

class Solution {
    fun validTree(n: Int, edges: Array<IntArray>): Boolean {
        // build 2-directional graph
        // check for cycle
        // check all nodes are connected

        val graph = buildGraph(edges)
        val visited = HashSet<Int>()

        val hasCycle = hasCycle(graph, visited)

        println("graph=$graph, hasCycle=$hasCycle, visited=$visited")

        return !hasCycle && visited.size == n
    }

    private fun hasCycle(
        graph: Map<Int, List<Int>>,
        visited: HashSet<Int>
    ): Boolean {
        // [[0,1],[1,3],[3,0],[2,4]]
        // 0 -> 1 -> 3
        // 0   <-    3
        //

        val root = 0
        val visitedFrom = HashMap<Int, Int?>()

        fun hasCycleDfs(node: Int, from: Int?): Boolean {
            val nextNodes = graph[node] ?: emptyList()

            visited.add(node)

            val prevFrom = visitedFrom[node]
            if (visitedFrom.contains(node) && prevFrom != from) return true

            visitedFrom[node] = from

            for (nextNode in nextNodes) {
                if (nextNode != from && hasCycleDfs(nextNode, node)) return true
            }

            return false
        }

        return hasCycleDfs(root, null)
    }

    private fun buildGraph(edges: Array<IntArray>): Map<Int, List<Int>> {
        val graph = HashMap<Int, MutableList<Int>>()

        for (edge in edges) {
            val (from, to) = edge
            graph[from] = graph.getOrDefault(from, mutableListOf())
                .apply {
                    add(to)
                }
            graph[to] = graph.getOrDefault(to, mutableListOf())
                .apply {
                    add(from)
                }
        }

        return graph
    }
}

